### DISCUSSION TITLE
Enabling TCP Connection with Nginx Stream Snippets

### DISCUSSION THREAD

**Discussion by robincher** [nginx/kubernetes-ingress#2011] [Q&A]
Hi Team,

I trying to expose some (fluentd) TCP services, and decided to use stream snippets to implement it. I have some issues configuring it with GlobalConfigurations and TransportServer CRDs due to my cluster set-up,but i leave it to another discussion.

Just want to check how i can properly/correctly pass in the stream snippets values via Helm.

Sample values.yaml snippets

```
 controller:
      config:
        name: nginx-config
        entries: 
          stream-snippets: |
            upstream fluentd-nginx-parser {
              server fluentd.test.svc.cluster.local:24221;
            }
            server {
                listen 24221 tcp;
                proxy_pass fluentd-nginx-parser;
            }
            upstream fluentd-general-parser {
                server fluentd.test.svc.cluster.local:24220;
            }
            server {
                listen 24220 tcp;
                proxy_pass fluentd-general-parser;
            } 
      globalConfiguration:
        create: false
      enableSnippets: true
```

Error from the controller

```
stderr: "2021/09/27 08:50:30 [emerg] 114#114: host not found in upstream \"fluentd.test.svc.cluster.local:24221\" in /etc/nginx/nginx.conf:142\nnginx: [emerg] host not found in upstream \"fluentd.test.svc.cluster.local:24221\" in /etc/nginx/nginx.conf:142\n"
finished with error: exit status 1
```

Do i need to pass in additional stuffs? Or i need to reload some stuffs.

I try running telnet from the nginx namespace to the service, and its working fine

```
kubectl run telnet-pod -it --rm --image:busybox -- /bin/sh -c 'telnet fluentd.test.svc.cluster.local 24221'
```

Thanks!

**Comment by jasonwilliams14** [comment]
@robincher 
can you provide the output of `nginx -T` ?
I'd like to see the NGINX configuration that is generated..
Thank you.

**Comment by robincher** [comment]
Hi @jasonwilliams14 ,

Not to hold you up, but i have decided to redeploy it with the community ingress nginx controller for now.

I revisit this thread when i revert back to use nginxinc ingress. Appreciate your help by the way.

**Comment by jasonwilliams14** [comment]
@robincher ok.
i was able to get your config working in my lab (adjustments to fit my specific lab) but it should work. 
reviewing the config would be helpful to see what is generated and what might need to be adjusted and tweaked

**Comment by BloodyIron** [comment]
@robincher 
Can you provide a sanitised example (full yaml manifest) of your ingress on what worked with the community NGINX Ingress? I'm trying to do this now, and I would like to see your example.

@jasonwilliams14 can I see your example too please? :)

I am unsure if I need to declare a spec.rules.host as part of this, and other such details, but this exact method I am exploring for similar TCP functions. I am also insure if I need to make one ingress per host involved, as the traffic routing requires it to be routed based on FQDN, as the application receiving it is expected to only respond to traffic destined for the FQDN it uses/expects, and nothing else (on the TCP aspects).
